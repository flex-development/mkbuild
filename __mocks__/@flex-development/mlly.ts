/**
 * @file Mocks - @flex-development/mlly
 * @module mocks/flex-development/mlly
 * @see https://github.com/flex-development/mlly
 */

/**
 * [`@flex-development/mlly`][1] module type.
 *
 * [1]: https://github.com/flex-development/mlly
 */
type Actual = typeof import('@flex-development/mlly')

/**
 * `@flex-development/mlly` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/mlly')

export const CONDITIONS = actual.CONDITIONS
export const RESOLVE_EXTENSIONS = actual.RESOLVE_EXTENSIONS
export const fillModules = vi.fn(actual.fillModules)
export const findStaticImports = vi.fn(actual.findStaticImports)
export const getSource = vi.fn(actual.getSource)
export const readPackageJson = vi.fn(actual.readPackageJson)
export const resolveModule = vi.fn(actual.resolveModule)
export const resolveModules = vi.fn(actual.resolveModules)
export const toDataURL = vi.fn(actual.toDataURL)
export const toURL = vi.fn(actual.toURL)
