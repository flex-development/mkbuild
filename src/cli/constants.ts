/**
 * @file CLI - Constants
 * @module mkbuild/cli/constants
 */

import pkg from '#pkg' assert { type: 'json' }

/**
 * Array containing choices for boolean flags.
 *
 * @internal
 *
 * @const {string[]} CHOICES_BOOLEAN
 */
export const CHOICES_BOOLEAN: string[] = ['false', 'true', '0', '1']

/**
 * CLI application name.
 *
 * @internal
 *
 * @const {string} CLI_NAME
 */
export const CLI_NAME: string = pkg.name.replace(/.*\//, '')
