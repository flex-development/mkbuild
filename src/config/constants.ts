/**
 * @file Config - Constants
 * @module mkbuild/config/constants
 */

/**
 * Default ignore patterns.
 *
 * @see https://github.com/mrmlnc/fast-glob#ignore
 *
 * @const {string[]} IGNORE_PATTERNS
 */
export const IGNORE_PATTERNS: string[] = [
  '**/.DS_Store',
  '**/.npmignore',
  '**/.yarnignore',
  '**/**.mdx',
  '**/**.stories.{j,t}sx',
  '**/{__mocks__,__snapshots__,__tests__}/**'
]

/**
 * Module extensions checked when attempting to resolve path aliases and module
 * specifiers.
 *
 * @const {string[]} MODULE_EXTENSIONS
 */
export const MODULE_EXTENSIONS: string[] = [
  '.cjs',
  '.cts',
  '.js',
  '.json',
  '.jsx',
  '.mdx',
  '.mjs',
  '.mts',
  '.ts',
  '.tsx'
]

/**
 * `require` and `require.resolve` statement regex.
 *
 * @todo prevent matching non-statements (e.g. comments, arbitrary strings)
 *
 * @const {RegExp} REQUIRE_REGEX
 */
export const REQUIRE_REGEX: RegExp =
  /(?:require\(|require\.resolve\()["']([^"']*)["']\)/g
