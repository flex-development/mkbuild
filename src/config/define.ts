/**
 * @file Config - defineBuildConfig
 * @module mkbuild/config/defineBuildConfig
 */

import type { Config } from '#src/interfaces'

/**
 * Helper for defining a build configuration object.
 *
 * @param {Config} [config={}] - Build configuration options
 * @return {Config} `config`
 */
const defineBuildConfig = (config: Config = {}): Config => config

export default defineBuildConfig
