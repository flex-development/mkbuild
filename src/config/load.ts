/**
 * @file Config - loadBuildConfig
 * @module mkbuild/config/loadBuildConfig
 */

import type { Config } from '#src/interfaces'
import { cosmiconfig } from 'cosmiconfig'
import es from './loader-es'

/**
 * Searches for a build configuration file.
 *
 * If found, a config object will be returned. The object will contain options
 * from the config file. If not found, an empty object will be returned.
 *
 * @see https://github.com/davidtheclark/cosmiconfig#explorersearch
 *
 * @async
 *
 * @param {string} [location] - Directory to search
 * @return {Config} Build configuration options
 */
const loadBuildConfig = async (location: string): Promise<Config> => {
  /**
   * Module name.
   *
   * @const {string} name
   */
  const name: string = 'build'

  // get search function
  const { search } = cosmiconfig(name, {
    loaders: { '.cts': es, '.js': es, '.mjs': es, '.mts': es, '.ts': es },
    searchPlaces: [
      '#.config.cjs',
      '#.config.cts',
      '#.config.js',
      '#.config.json',
      '#.config.mjs',
      '#.config.mts',
      '#.config.ts'
    ].map(place => place.replace('#', name)),
    stopDir: location
  })

  return ((await search(location))?.config as Config | null) ?? {}
}

export default loadBuildConfig
