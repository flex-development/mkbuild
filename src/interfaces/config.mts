/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystem, Task } from '@flex-development/mkbuild'

/**
 * Build configuration.
 *
 * @see {@linkcode Task}
 *
 * @extends {Task}
 */
interface Config extends Task {
  /**
   * File system API.
   *
   * @see {@linkcode FileSystem}
   */
  fs?: Partial<FileSystem> | null | undefined

  /**
   * Build tasks.
   *
   * Each build task will be merged with the base task. Top-level lists (arrays
   * and sets) are concatted and deduped. Plain objects are deep merged.
   *
   * > ðŸ‘‰ **Note**: If empty, `null`, or `undefined`, a single build task will
   * > be inferred from the base task.
   *
   * @see {@linkcode Task}
   */
  tasks?: Task[] | null | undefined
}

export type { Config as default }
