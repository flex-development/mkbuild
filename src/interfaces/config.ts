/**
 * @file Interfaces - Config
 * @module mkbuild/interfaces/Config
 */

import type { FileSystemAdapter } from '#src/types'
import type * as esbuild from 'esbuild'
import type Options from './options'
import type Task from './task'

/**
 * Build configuration options.
 *
 * @extends {Options}
 */
interface Config extends Options {
  /**
   * Load build config from config file.
   *
   * @default true
   */
  configfile?: boolean

  /**
   * Build entries.
   *
   * **Note**: If empty or `undefined`, a single build entry will be inferred
   * from the remaining set of configuration options.
   *
   * @default []
   */
  entries?: Partial<Omit<Task, 'write'>>[]

  /**
   * Custom implementations of file system methods.
   *
   * @see {@linkcode FileSystemAdapter}
   * @see https://nodejs.org/api/fs.html
   *
   * @default fsa
   */
  fs?: FileSystemAdapter

  /**
   * Serve files.
   *
   * @see {@linkcode esbuild.ServeOptions}
   * @see https://esbuild.github.io/api/#serve
   *
   * @default false
   */
  serve?: esbuild.ServeOptions | boolean

  /**
   * Watch files.
   *
   * @default false
   */
  watch?: boolean

  /**
   * Write build results.
   *
   * @default false
   */
  write?: boolean
}

export type { Config as default }
