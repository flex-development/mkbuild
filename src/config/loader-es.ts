/**
 * @file Config - esLoader
 * @module mkbuild/config/esLoader
 */

import type { Config } from '#src/interfaces'
import tsconfigPaths from '#src/plugins/tsconfig-paths/plugin'
import { build } from 'esbuild'
import { resolveImports, toDataURL } from 'mlly'
import { RESOLVE_EXTENSIONS } from './constants'

/**
 * Loads an [ESM][1] (`*.js`, `*.mjs`) or TypeScript (`*.cts`, `*.mts`, `*.ts`)
 * build config file.
 *
 * [1]: https://nodejs.org/api/esm.html
 *
 * @see https://github.com/davidtheclark/cosmiconfig#loaders
 *
 * @async
 *
 * @param {string} path - Config file path
 * @param {string} content - Config file content
 * @return {Promise<Config>} Build configuration options
 */
const esLoader = async (path: string, content: string): Promise<Config> => {
  const {
    outputFiles: [output]
  } = await build({
    entryPoints: [path],
    format: 'esm',
    loader: {
      '.cts': 'ts',
      '.js': 'js',
      '.mjs': 'js',
      '.mts': 'ts',
      '.ts': 'ts'
    },
    metafile: true,
    outdir: '<stdout>',
    plugins: [tsconfigPaths()],
    write: false
  })

  // resolve imports in content to evaluate module
  content = await resolveImports(output!.text, {
    extensions: RESOLVE_EXTENSIONS,
    url: path
  })

  return ((await import(toDataURL(content))) as { default: Config }).default
}

export default esLoader
