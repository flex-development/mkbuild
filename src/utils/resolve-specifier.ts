/**
 * @file Utilities - resolveSpecifier
 * @module mkbuild/utils/resolveSpecifier
 */

import { resolvePath } from 'mlly'
import * as pathe from 'pathe'
import { MODULE_EXTENSIONS } from 'src/config/constants'
import type { BuildEntry, OutputFile, Statement } from 'src/interfaces'

/**
 * Resolves a relative specifier in `statement.code` and `statement.specifier`.
 *
 * A relative specifier, like `'./startup.js'` or `'../config.mjs'`, refers to a
 * path relative to the location of the importing file.
 *
 * Unless `NODE_OPTIONS` is being used to control specifier resolution, the file
 * extension is always necessary when using esm.
 *
 * @see https://nodejs.org/docs/latest-v16.x/api/esm.html#terminology
 *
 * @async
 *
 * @param {Pick<Statement, 'code' | 'specifier'>} statement - Statement object
 * @param {Pick<BuildEntry, 'ext' | 'format'>} entry - Build entry
 * @param {BuildEntry['ext']} [entry.ext] - Build entry extension
 * @param {BuildEntry['format']} [entry.format] - Build entry format
 * @param {OutputFile['src']} src - Full path to output source file
 * @return {Promise<void>} Nothing when complete
 */
const resolveSpecifier = async (
  statement: Pick<Statement, 'code' | 'specifier'>,
  entry: Pick<BuildEntry, 'ext' | 'format'>,
  src: OutputFile['src']
): Promise<void> => {
  /**
   * {@link statement} specifier before module path resolution.
   *
   * @const {string | undefined} specifier
   */
  const specifier: string | undefined = statement.specifier

  // do nothing if missing entry extension, entry format, or module specifier
  if (!entry.ext || !entry.format || !specifier) return

  // do nothing if specifier is not relative or already has extension
  if (!specifier.startsWith('.') || pathe.extname(specifier)) return

  /**
   * {@link statement.specifier} resolved.
   *
   * @see https://github.com/unjs/mlly#resolvepath
   *
   * @const {string} resolved
   */
  const resolved: string = await resolvePath(specifier, {
    conditions: [entry.format === 'esm' ? 'import' : 'require'],
    extensions: MODULE_EXTENSIONS,
    url: src
  })

  const { name } = pathe.parse(specifier)
  const { name: resolved_name } = pathe.parse(resolved)

  // specifier resolved to a directory
  if (name !== resolved_name) statement.specifier += '/' + resolved_name

  // add build entry extension to specifier
  statement.specifier += entry.ext

  // reset statement code
  statement.code = statement.code.replace(specifier, statement.specifier!)

  return void statement
}

export default resolveSpecifier
