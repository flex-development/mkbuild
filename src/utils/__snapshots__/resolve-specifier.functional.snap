// Vitest Snapshot v1

exports[`functional:utils/resolveSpecifier > resolve > relative > should resolve relative path to directory 1`] = `
{
  "code": "export * from './interfaces/index.mjs'",
  "end": 38,
  "specifier": "./interfaces/index.mjs",
  "start": 0,
}
`;

exports[`functional:utils/resolveSpecifier > resolve > relative > should resolve relative path to file 1`] = `
{
  "code": "const { REQUIRE_REGEX } = require('../config/constants.cjs')",
  "end": 60,
  "specifier": "../config/constants.cjs",
  "start": 0,
}
`;
